sql/settype.go

```diff
@@ -0,0 +1,359 @@
+}
+
+// String implements Type interface.
+func (t setType) String() string {
+	s := fmt.Sprintf("SET('%v')", strings.Join(t.Values(), ","))
+	if t.CharacterSet() != Collation_Default.CharacterSet() {
+		s += " CHARACTER SET " + t.CharacterSet().String()
+	}
+	if t.collation != Collation_Default {
+		s += " COLLATE " + t.collation.String()
+	}
+	return s
+}
+
+// Type implements Type interface.
+func (t setType) Type() query.Type {
+	return sqltypes.Set
+}
+
+// Zero implements Type interface.
+func (t setType) Zero() interface{} {
+	return ""
+}
+
+func (t setType) CharacterSet() CharacterSet {
+	return t.collation.CharacterSet()
+}
+
+func (t setType) Collation() Collation {
+	return t.collation
+}
+
+// ConvertBitArrayToString converts the given bit array into the equivalent comma-delimited string.
+// The given bit array must have a form where representing the value 12 would have values {4,8}.
+func (t setType) ConvertBitArrayToString(bitArray []uint64) (string, error) {
+	strBuilder := strings.Builder{}
+	for i, bit := range bitArray {
+		val, ok := t.bitToVal[bit]
+		if !ok {
+			return "", ErrInvalidSetValue.New(t.sumBitArray(bitArray))
+		}
+		if i != 0 {
+			strBuilder.WriteByte(',')
+		}
+		strBuilder.WriteString(val)
+	}
+	return strBuilder.String(), nil
+}
+
+// ConvertStringToBitArray converts the given string into a bit array.
+// Ignores duplicate set values and always returns the bits in ascending order.
+// The returned bit array representing the value 12 will have values {4,8}.
+func (t setType) ConvertStringToBitArray(str string) ([]uint64, error) {
+	if str == "" {
+		return nil, nil
+	}
+	var bitArray []uint64
+	seen := make(map[string]struct{})
+	vals := strings.Split(str, ",")
+	for _, val := range vals {
+		if _, ok := seen[val]; ok {
+			continue
+		}
+		seen[val] = struct{}{}
+		var compareVal string
+		switch t.collation {
+		case Collation_binary:
+			compareVal = val
+		default:
+			compareVal = strings.ToLower(strings.TrimRight(val, " "))
+		}
+		if originalVal, ok := t.compareToOriginal[compareVal]; ok {
+			bitArray = append(bitArray, t.valToBit[originalVal])
+		} else {
+			return nil, ErrInvalidSetValue.New(val)
+		}
+	}
+	sort.Slice(bitArray, func(i, j int) bool {
+		return bitArray[i] < bitArray[j]
+	})
+	return bitArray, nil
+}
+
+// Values returns all of the set's values in ascending order according to their corresponding bit value.
+func (t setType) Values() []string {
+	// A set with 3 values will have an upper bound of 8, or 0b1000.
+	// 8 - 1 == 7, and 7 is 0b0111, which would map to every value in the set.
+	bitArray := t.decomposeToBitArray(t.upperBound() - 1)
+	valArray := make([]string, len(bitArray))
+	for i, bit := range bitArray {
+		valArray[i] = t.bitToVal[bit]
+	}
+	return valArray
+}
+
+// decomposeToBitArray returns an array of all of the bits that were set to 1.
+// For example, 12 will return {4,8} for 0b1100.
+func (t setType) decomposeToBitArray(num uint64) []uint64 {
```

Comment:
Same comment here. Just do the bit math by repeatedly shifting, no need to assemble an array of values when you can use a single int64.